pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -318,21 +381,83 @@ func (rd *restoreDataProcessor) openSSTs(
 		return nil
 	}
 
+	hashSpan := func(span roachpb.Span) (uint32, error) {
+		hash := fnv.New32a()
+		_, err := hash.Write(span.Key)
+		if err != nil {
+			return 0, err
+		}
+
+		return hash.Sum32() % uint32(rd.numWorkers), nil
+	}
+
 	log.VEventf(ctx, 1 /* level */, "ingesting span [%s-%s)", entry.Span.Key, entry.Span.EndKey)
 
-	storeFiles := make([]storageccl.StoreFile, 0, len(EntryFiles{}))
-	for _, file := range entry.Files {
+	storeFiles := make([]storageccl.StoreFile, 0, len(entry.Files))
+	var sstOverheadBytesPerFile int64
+	if rd.spec.Encryption != nil {
+		sstOverheadBytesPerFile = sstReaderEncryptedOverheadBytesPerFile
+	} else {
+		sstOverheadBytesPerFile = sstReaderOverheadBytesPerFile
+	}
+
+	for idx := 0; idx < len(entry.Files); {
+		file := entry.Files[idx]
 		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
 
+		if err := rd.mem.Grow(ctx, sstOverheadBytesPerFile); err != nil {
+			// If we failed to allocate more memory, send the iterator
+			// containing the files we have right now.
+			if len(storeFiles) > 0 {
+				iterOpts := storage.IterOptions{
+					RangeKeyMaskingBelow: rd.spec.RestoreTime,
+					KeyTypes:             storage.IterKeyTypePointsAndRanges,
+					LowerBound:           keys.LocalMax,
+					UpperBound:           keys.MaxKey,
+				}
+				iter, err := storageccl.ExternalSSTReader(ctx, storeFiles, rd.spec.Encryption, iterOpts)
+				if err != nil {
+					return err
+				}
+				workerNum, err := hashSpan(entry.Span)
+				if err != nil {
+					return err
+				}
+
+				log.VInfof(ctx, 2, "sending iterator after %d out of %d files due to insufficient memory", idx, len(entry.Files))
+				if err := sendIter(iter, dirs, workerSSTChannels[workerNum], int64(len(storeFiles))*sstReaderOverheadBytesPerFile, false); err != nil {
+					return err
+				}
+				storeFiles = storeFiles[:0]
+			}
+
+			waitCh := make(chan struct{})
+			rd.mu.Lock()
+			if rd.mu.memWaitCh != nil {
+				rd.mu.Unlock()
+				return errors.AssertionFailedf("there should only be 1 instance of openSSTs per restore data processor")
+			}
+			rd.mu.memWaitCh = waitCh
+			rd.mu.Unlock()
+
+			select {
+			case <-waitCh:
```

Comment:
I'm trying to figure out if we might still have a lock order acquisition problem here:
What happens if right after we fail to `Grow()` all the pending flushes finish and release, before we set memWaitCh, so they breeze past it. then let's say our `sendIter` work also happen to run to completion, including  before we get around to setting `memWaitCh` below, so by the time we set it, _nobody_ is out there with a pending SST and going to call shrink and then close memWaitCh for us when it finishes?

Reply:
I wonder if a single shared IntPool would be simpler here, instead of the mutex+ch? I haven't completely thought it through, so maybe not. In my head, it might look something like:  
* First, where we just `Grow()` today, we'd instead `TryAcquire` -- if that fails, we'd then try to `Grow()` and if that succeeds, update the pool capacity with the growth amount then go back to `TryAcquire`.
* If `TryAcquire` and `Grow()` both fail, we enter this block, send the iter, etc.
* Then we just do plain blocking Acquire() here, instead of mu/ch, and don't need the `continue` either: either via the TryAcquire or via this blocking Acquire, either way you reach the end of the block with an inc alloc reservation?

Reply:
done, now using a quota pool